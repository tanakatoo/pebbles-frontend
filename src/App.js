import './styles/app.css';
import NavBar from "./components/common/NavBar"
import "./styles/app.css"
import { useSelector } from 'react-redux';
import MyRoutes from "./MyRoutes"

function App() {
  //get the font from the store
  const font = useSelector(state => state.langFont.font)

  //we have to hardcode the string of the font because tailwind doesn't know what to use at build
  return (
    <div className={`App ${font == 'font-poppins' ? 'font-poppins' : 'font-notoJP'}`}>
      <NavBar />
      <MyRoutes />
    </div >
  );
}

export default App;
