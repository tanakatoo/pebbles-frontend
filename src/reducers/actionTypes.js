/* PROFILE action types */
export const LOAD_PROFILE = "loadProfile"
export const LOGOUT = "logout"
export const LOGIN = "login"

/* LANGUAGE action types */
export const CHANGE_LANG = "changeLang"
export const CHANGE_PAGE = "changePage"