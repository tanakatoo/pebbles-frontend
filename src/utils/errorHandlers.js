import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}



export const errorHandlers = [

    //login
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {
        const failLogin = {
            "error": {
                "message": "INVALID_CREDENTIALS",
                "status": 401
            }
        }
        console.log(res(ctx.json([failLogin])))
        return res(ctx.json([failLogin]));
    }),

    rest.post('http://localhost:3001/email', (req, res, ctx) => {
        console.log('MOCK error /email');
        const noData = {
            "error": {
                "message": "no data",
                "status": 400
            }
        }
        return res(ctx.json(noData));
    }),

]