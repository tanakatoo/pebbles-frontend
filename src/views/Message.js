import React, { useEffect, useRef, useState, forwardRef } from 'react'
import Protected from '../components/common/Protected'
import Card from '../components/message/Card'
import usePageText from '../hooks/usePageText'
import useSetToken from '../hooks/useSetToken'
import { v4 as uuid } from "uuid"
import MessageApi from '../api/message'
import { useParams } from 'react-router-dom'
import ServerError from '../components/form/ServerError'
import Spinner
    from '../components/common/Spinner'
import moment from 'moment-timezone'
import 'moment/locale/ja'
import ConversationTitle from '../components/message/ConversationTitle'
import { AwesomeSend } from '../styles/Icons'
import { useSelector } from 'react-redux'
import useFormData from '../hooks/useFormData'
import { Link } from 'react-router-dom'

function Message() {

    const INITIAL_DATA = {
        msg: ''
    }
    const [msg, setMsg, handleChange, resetMsg] = useFormData(INITIAL_DATA)
    const [doneGettingData, setDoneGettingData] = useState(false)
    const msgRef = useRef(null)
    const [data, setData] = useState([])
    const [errors, setErrors] = useState([])
    const [token = null] = useSetToken()
    const [pageText, lang] = usePageText("messages")
    const { username } = useParams()
    const user = useSelector(state => state.profile.profile.username)

    const getData = async () => {
        try {
            setErrors([])
            setDoneGettingData(false)
            //makes data all be read when getting data
            const res = await MessageApi.getConversation(username)
            console.log(res)
            setData(res)
            setDoneGettingData(true)
        } catch (e) {
            if (e instanceof TypeError) {
                //means server is down
                console.error('Type error in getting data', e)
                setErrors(["UNKNOWN"])
            } else {
                console.log(e[0] == 'blocked')
                if (e[0] == 'blocked') {
                    console.log('in setting errors blocked')
                    setErrors(["BLOCKED"])
                } else {
                    console.error('Error getting data', e)
                    setErrors(e)
                }

            }
        }
    }
    useEffect(() => {
        getData()
    }, [])

    useEffect(() => {
        if (msgRef.current) {
            msgRef.current.style.height = '0px';
            const scrollHeight = msgRef.current.scrollHeight;
            console.log('scrollheight is', scrollHeight)
            msgRef.current.style.height = `${scrollHeight}px`;
        }

    }, [msgRef.current, msg])

    let lastSent
    let conversationWithUsername
    let conversationWithAvatar
    if (data.length > 0) {
        conversationWithUsername = user === data[0].from ? data[0].to : data[0].from
        conversationWithAvatar = user === data[0].from ? data[0].toavatar : data[0].fromavatar
        lastSent = moment.utc(data[0].sent_at, 'YYYY-MM-DD HH:mm:ss').tz(Intl.DateTimeFormat().resolvedOptions().timeZone).locale(lang.toLowerCase()).format('LLL')

    }

    const handleSubmit = async () => {
        setErrors([])
        try {
            console.log('data to send', msg.msg)
            const res = await MessageApi.sendMsg(username, msg.msg)
            getData()

        } catch (e) {
            if (e instanceof TypeError) {
                //means server is down
                console.log('Type error in submit', e)
                setErrors(["UNKNOWN"])
            } else {
                console.log('Error in submit', e)
                setErrors(e)
            }
        } finally {
            resetMsg()

        }
    }


    return (
        // <Protected>
        <div className=' flex flex-col justify-center items-center'>
            {errors.length > 0 &&
                <div>
                    <ServerError msg={errors} />
                    {errors == "BLOCKED" ?
                        <p className='text-center mb-12'>
                            <Link className='underline' to="/users/unblock">
                                {pageText.UNBLOCK_CONTACTS}</Link></p> : ''}

                </div>}
            {data.length === 0 && errors.length === 0 && doneGettingData === false ?
                <div className='my-24'><Spinner /></div>
                : errors.length === 0 && data.length > 1 || doneGettingData === true ?
                    <>
                        <ConversationTitle link={`/users/${conversationWithUsername}`}
                            src={`../avatars/${conversationWithAvatar}`}
                            username={`${conversationWithUsername}`} />
                        <div className='w-full max-w-[500px] bg-background shadow-sm pb-12'>

                            <form className='w-full flex items-center px-2 my-2'>
                                <div className='flex flex-col w-full m'>
                                    <textarea ref={msgRef} id='msg' name='msg' value={msg.msg} onChange={handleChange} rows="1"
                                        className={`overflow-hidden mb-2 rounded-ml py-3 px-4 text-black placeholder-gray`} />
                                </div>
                                <div className='text-primary flex mx-2 cursor-pointer'><AwesomeSend size="xl" onClick={handleSubmit} /></div>

                            </form>

                            <div className='w-full grid grid-cols-4 gap-x-4 md:grid-cols-8 lg:grid-cols-12  
                                                lg:max-w-[900px] content-center '>
                                {/* <div className='col-span-full mb-4 mt-2'><SearchBar handleSearch={handleSearch} /></div> */}
                                <div className='col-span-full'>
                                    {data[0].fromavatar !== '' ?
                                        data.map((d, idx) => (
                                            <React.Fragment key={uuid()}>
                                                <Card data={d} latest={idx === 0 ? true : false} />
                                                {idx === 0 ? <>
                                                    <div className={`flex ${user === data[0].from ? 'justify-end' : 'justify-start'} mx-4 mb-3`}>
                                                        <p className='text-gray text-mobile-label-2'>{pageText.LAST_SENT} {lastSent}</p>
                                                    </div>
                                                </>
                                                    : ''
                                                }
                                            </React.Fragment>)


                                        )

                                        : ''}

                                </div>
                            </div>
                        </div>
                    </>
                    : ''
            }

        </div >
        // </Protected >
    )
}

export default Message