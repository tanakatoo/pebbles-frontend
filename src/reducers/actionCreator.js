import axios from "axios"
import useLocalStorage from "../hooks/useLocalStorage"
import { LOGIN, LOGOUT } from "./actionTypes"
import AuthApi from "../api/auth";

export function actionLogin(username, password) {
    return async function (dispatch) {
        const res = await AuthApi.login(username, password)
        dispatch(actionLoggedIn(res))
    }
}

function actionLoggedIn(token) {
    return ({ type: LOGIN, token: token })
}


// export function actionLogout(token) {
//     return async function (dispatch) {
//         const [storageToken, setStorageToken, removeLocalStorage] = useLocalStorage("token")

//         //set the token state to null
//         setStorageToken(null)
//         //remove the token from the storage
//         removeLocalStorage()
//         dispatch(actionLoggedOut())
//     }
// }

// function actionLoggedOut() {
//     return ({ type: LOGOUT })
// }
