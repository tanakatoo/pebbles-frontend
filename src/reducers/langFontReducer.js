import { CHANGE_LANG, CHANGE_PAGE } from "./actionTypes"
import language from '../text/language.json';

console.log('navigator', navigator.language)
let initial_lang = "JA"
if (/^en\b/.test(navigator.language)) {
    console.log('ther eis en in the lang')
    initial_lang = "EN"
}

const INITIAL_STATE = { lang: initial_lang, font: language.FONTS[initial_lang], allFonts: language.FONTS }

const langFontReducer = (state = INITIAL_STATE, action) => {
    switch (action.type) {
        case CHANGE_LANG:
            //changes the language and sets the appropraite font and text 
            return { ...state, lang: action.lang, font: language.FONTS[action.lang] }

        default:
            return state

    }
}

export default langFontReducer
