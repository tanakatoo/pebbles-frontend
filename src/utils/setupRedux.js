import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
// As a basic setup, import your same slice reducers
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import { PersistGate } from 'redux-persist/integration/react'
import { ListSlice } from "../reducers/rootReducer";
import rootReducer from "../reducers/rootReducer"
import storage from 'redux-persist/lib/storage'
import storeSetup from "./storeSetup"
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

// const persistConfig = {
//     key: 'root', // Key prefix for the persisted state
//     storage, // Storage mechanism (e.g., local storage)
//     stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
//     whitelist: ['profile', 'langFont']
// };
// const persistedReducer = persistReducer(persistConfig, rootReducer);

// const persistedStore = persistStore(store)
// const root = ReactDOM.createRoot(document.getElementById('root'));


export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = storeSetup(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {

        return <Provider store={store}>{children}</Provider>;
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}