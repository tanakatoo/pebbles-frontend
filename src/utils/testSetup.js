// import { persistedStore, store } from './setupRedux';
import { render } from '@testing-library/react'
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react'
import { setupTestStore } from './setupRedux';


const renderWithProviders = (ui, options = {}) => {
    const [store, persistedStore] = setupTestStore(options.preloadedState);

    const AllProviders = ({ children }) => {
        return (
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistedStore}>
                    {children}
                </PersistGate>
            </Provider>
        )
    };

    return { store, ...render(ui, { wrapper: AllProviders, ...options }) };
}


// "re-export everything"
export * from '@testing-library/react';

export { renderWithProviders };