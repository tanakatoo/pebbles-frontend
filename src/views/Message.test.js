import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, renderWithProviders } from '../utils/testSetup';
import Message from './Message';
import App from '../App';
import { waitFor } from '../utils/testSetup';
import { act } from 'react-dom/test-utils';
import { server } from '../utils/server'
import { errorHandlers } from '../utils/errorHandlers';
import { actionLogin } from '../reducers/actionCreator';
import { useDispatch } from 'react-redux';
import { store } from '../utils/setupRedux';
import { LOGIN } from '../reducers/actionTypes';

// Mock window.scrollTo()
window.scrollTo = jest.fn();
afterAll(() => {
    jest.clearAllMocks();
});
let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}

beforeEach(async () => {
    // const dispatch = useDispatch()

    const { getByTestId, getByLabelText, getByText } = renderWithProviders(
        <MemoryRouter >
            <App />
        </MemoryRouter>
    );

    // Populate form
    let changeLang = getByText('EN', { exact: true });
    fireEvent.click(changeLang);

    //login first - this one doesn't work
    await act(() => {
        store.dispatch({
            type: LOGIN,

            token: 'abced',
            profile: {
                id: 2,
                username: 'hello',
                role: 'regular',
                premium_join_date: '',
                premium_end_date: '',
                free_trial_start_date: '',

            }
        })
    })


})

afterEach(() => {
    window.localStorage.token = null
})


describe('Message', () => {
    test('renders without crashing', async () => {
        await act(() => {
            renderWithProviders(
                <MemoryRouter>
                    <Message />
                </MemoryRouter>
                , {
                    preloadedState: {
                        profile: {
                            token: 'abced',
                            profile: loginResponse
                        }
                    }
                });
        })

    });

    test('matches snapshot', async () => {
        await act(() => {
            const { asFragment } = renderWithProviders(
                <MemoryRouter>
                    <Message />
                </MemoryRouter>
            );
            expect(asFragment()).toMatchSnapshot();
        })


    });


    test('add to conversation', async () => {
        const { getByTestId, getByRole, getByLabelText, getByText } = await renderWithProviders(
            <MemoryRouter >
                <Message />
            </MemoryRouter >
        );


        let elem = getByRole('textbox');
        fireEvent.change(elem, { target: { value: 'hi' } });
        fireEvent.keyDown(elem, { key: 'Enter', code: 'Enter', charCode: 13 })


        await waitFor(() => {
            expect(getByText(/Last sent/)).toBeInTheDocument();
            expect(getByText(/hi/)).toBeInTheDocument();
        });

    });


});