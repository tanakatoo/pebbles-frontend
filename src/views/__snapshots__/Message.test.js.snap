// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message matches snapshot 1`] = `<DocumentFragment />`;
